lib/util.c

```diff
@@ -312,3 +312,39 @@ doubles_almost_equal(double a, double b, double eps)
 {
     return (fabs(a) < eps && fabs(b) < eps) || gsl_fcmp(a, b, eps) == 0;
 }
+
+/* Inputs:
+ *   - `lengths` is an array of the real number lengths of `num_intervals`
+ *      consecutive half-open intervals on the real number line start at zero,
+ *      i.e. given lengths L_i, the intervals are:
+ *          [0.0, L_0), [L_0, L_0 + L_1), [L_0 + L_1, L_0 + L_1 + L_2), ...
+ *   - `x` is a real number to find in one of the intervals or after all of them
+ * Returns:
+ *   - the index (starting at zero) of the interval in which `x` was found
+ *   - `num_intervals` if `x` is at or past the end of the last interval
+ *     (the sum of all the lengths)
+ *   - zero if `x` is negative
+ *   - `num_interval` if `x` is NaN
+ * Pre-conditions:
+ *   - `lengths` must point to `num_interval` non-negative non-NaN doubles
+ * Post-condition:
+ *   - returned value is less than or equal to `num_intervals` (and non-negative)
+ *
+ */
+size_t
+positive_interval_select(double x, size_t num_intervals, double const *lengths)
+{
+    size_t i = 0;
+    double sum = 0.0;
+    while (i < num_intervals) {
+        assert(lengths[i] >= 0.0);
+        sum += lengths[i];
+        if (x < sum)
```

Comment:
We always use braces:
```c
if (x < sum) {
    break;
}
```

Reply:
OK. I'm trying to recall some of the gotchas when omitting braces. I think merge issues is one of the dangers. Does a particular bug-inducing danger in omitting braces come to mind?

Reply:
Yep - forgetting about them is absolutely the main issue here. Especially for people used to Python where indentation is meaningful, stuff like this is absolutely lethal:
```c
if (x < sum) 
    x++;
    y++;
y++;
```

Reply:
One could argue that this isn't a danger with a ``break``, but why bother? It's much simpler to pick one rule and to stick with it rather than wasting brain cycles on choosing the appropriate thing based on the situation.

Reply:
Good point on the Python programmers!

I'm going to waste some brain cycles trying to find a merge scenario where my code results in a bug. :) I have this hunch it's something one could call "merge unfriendly".

Not that I really care, I'm more just curious, like a field linguist of the "languages" "spoken" by programmers. :) But some, like those implementing clang, think the simple rule to just follow is NOT put braces with single simple statements. I suspect this might be part of the reason clang-format does not have a feature for forcing braces after if statements:
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

Reply:
Yeah, I think it all comes down to what you were brought up on essentially. I was taught to always use braces and it seemed like a good idea so I did it. Others disagree - no coding convention is perfect.

Reply:
I've been bitten by the situation where you come along some months later to add a line to the loop, forgetting to add the braces, donuts always been my preference to always use braces.